cmd/allocator/main.go

```diff
@@ -0,0 +1,165 @@
+	tlsDir  = "/home/allocator/tls/"
+	port    = "8443"
+)
+
+// A handler for the web server
+type handler func(w http.ResponseWriter, r *http.Request)
+
+func main() {
+	agonesClient, err := getAgonesClient()
+	if err != nil {
+		logger.WithError(err).Fatal("could not create agones client")
+	}
+
+	h := httpHandler{
+		agonesClient: agonesClient,
+		namespace:    os.Getenv("NAMESPACE"),
+	}
+
+	// TODO: add liveness probe
+	http.HandleFunc("/v1alpha1/gameserverallocation", h.postOnly(h.allocateHandler))
+
+	caCertPool, err := getCaCertPool(certDir)
+	if err != nil {
+		logger.WithError(err).Fatal("could not get CA certs")
+	}
+
+	cfg := &tls.Config{
+		ClientAuth: tls.RequireAndVerifyClientCert,
+		ClientCAs:  caCertPool,
+	}
+	srv := &http.Server{
+		Addr:      ":" + port,
+		TLSConfig: cfg,
+	}
+
+	err = srv.ListenAndServeTLS(tlsDir+"tls.crt", tlsDir+"tls.key")
+	logger.WithError(err).Fatal("allocation service crashed")
+}
+
+// Set up our client which we will use to call the API
+func getAgonesClient() (*versioned.Clientset, error) {
+	// Create the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, errors.New("Could not create in cluster config")
+	}
+
+	// Access to the Agones resources through the Agones Clientset
+	agonesClient, err := versioned.NewForConfig(config)
+	if err != nil {
+		return nil, errors.New("Could not create the agones api clientset")
+	}
+
+	return agonesClient, nil
+}
+
+func getCaCertPool(path string) (*x509.CertPool, error) {
+	// Add all certificates under client-certs path because there could be multiple clusters
+	// and all client certs should be added.
+	caCertPool := x509.NewCertPool()
+	filesInfo, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, fmt.Errorf("error reading certs from dir %s: %s", path, err.Error())
+	}
+
+	for _, file := range filesInfo {
+		if strings.HasSuffix(file.Name(), ".crt") || strings.HasSuffix(file.Name(), ".pem") {
+			certFile := path + file.Name()
+			caCert, err := ioutil.ReadFile(certFile)
+			if err != nil {
+				return nil, fmt.Errorf("ca cert is not readable or missing: %s", err.Error())
+			}
+			if !caCertPool.AppendCertsFromPEM(caCert) {
+				return nil, fmt.Errorf("client cert %s cannot be installed", certFile)
+			}
+			logger.Infof("client cert %s is installed", certFile)
+		}
+	}
+
+	return caCertPool, nil
+}
+
+// Limit verbs the web server handles
+func (h *httpHandler) postOnly(in handler) handler {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == "POST" {
+			in(w, r)
+			return
+		}
+		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
+	}
+}
+
+type httpHandler struct {
+	agonesClient versioned.Interface
+	namespace    string
+}
+
+func (h *httpHandler) allocateHandler(w http.ResponseWriter, r *http.Request) {
+	gsa := allocationv1alpha1.GameServerAllocation{}
+	if err := json.NewDecoder(r.Body).Decode(&gsa); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```

Comment:
probably better not to return err.Error() to the user - just return something vague like "invalid request"